File: Assets/Plugins/iOS/CrossDKBridge.m
//
//  CrossDKBridge.m
//  CrossDKUnityBridge
//
//  Created by Adikteev on 29/03/2022.
//

#import "CrossDKBridge.h"

@interface CrossDKOverlayDelegate : NSObject<CrossDKOverlayDelegate>
@property (nonatomic, weak) CrossDKBridge *parentBridge;
- (instancetype)initWithParentBridge:(CrossDKBridge *)parentBridge;
@end

@interface CrossDKBridge ()
@property (nonatomic, strong) CrossDKOverlay* crossDKOverlay;
@property (nonatomic, strong) CrossDKOverlayDelegate *crossDKOverlayDelegate;
@end

@implementation CrossDKBridge

NSString* makeNSString (const char* string) {
    if (string) {
        return [NSString stringWithUTF8String: string];
    } else {
        return [NSString stringWithUTF8String: ""];
    }
}

char* makeCString(NSString *str) {
    const char* string = [str UTF8String];
    if (string == NULL) {
        return NULL;
    }

    char *buffer = (char*)malloc(strlen(string) + 1);
    strcpy(buffer, string);
    return buffer;
}

void unitySendMessage(const char *method, NSString *message) {
    NSLog(@"%@", message);
    UnitySendMessage("CrossDK", method, makeCString(message));
}

/// Sets up configuration for `CrossDK`.
///
/// - Parameters:
///     - appId: current application's App Store ID
///     - apiKey: authorization API key
///     - userId: user's ID
void crossDKConfigWithAppId(const char *appId, const char *apiKey, const char *userId) {
    NSString *nsAppID = makeNSString(appId);
    NSString *nsApiKey = makeNSString(apiKey);
    NSString *nsUserId = makeNSString(userId);
    [
        CrossDKConfig.shared
        setupWithAppId:nsAppID
        apiKey:nsApiKey
        userId:nsUserId
    ];
}

/// Displays an Overlay view.
///
/// - Parameters:
///     - format: banner, mid_size, interstitial overlay
///     - position: banner and mid_size overlay position
///     - withCloseButton: mid_size and interstitial overlay close button
///     - isRewarded: provides some kind of value for the user (interstitial format only)
 - (void) displayOverlayWithFormat:(OverlayFormat)isFormat position:(OverlayPosition)isPosition withCloseButton:(BOOL*)isWithCloseButton isRewarded:(BOOL*)isIsRewarded {
     _crossDKOverlay = [[CrossDKOverlay alloc] init];
     UIWindow *window = [[UIApplication sharedApplication] keyWindow];
     self.crossDKOverlayDelegate = [[CrossDKOverlayDelegate alloc] initWithParentBridge: self];
     self.crossDKOverlay.delegate = self.crossDKOverlayDelegate;

     if (window != nil) {
         [
             self.crossDKOverlay
             displayWithWindow:window
             format:isFormat
             position:isPosition
             withCloseButton:isWithCloseButton
             isRewarded:isIsRewarded
         ];
     }
 }

/// Dismisses an overlay view.
- (void) dismissOverlay {
    _crossDKOverlay = [[CrossDKOverlay alloc] init];
    UIWindow *window = [[UIApplication sharedApplication] keyWindow];
    if (window != nil) {
        [
            self.crossDKOverlay
            dismissWithWindow:window
        ];
    }
}

@end

static CrossDKBridge* delegateObject = nil;

/// Dismisses an overlay view.
void dismissOverlay() {
    if (delegateObject == nil)
        delegateObject = [[CrossDKBridge alloc] init];

    [delegateObject dismissOverlay];
}

/// Displays an Overlay view.
///
/// - Parameters:
///     - format: banner, mid_size, interstitial overlay
///     - position: banner and mid_size overlay position
///     - withCloseButton: mid_size and interstitial overlay close button
///     - isRewarded: provides some kind of value for the user (interstitial format only)
 void displayOverlayWithFormat(OverlayFormat format, OverlayPosition position, BOOL* withCloseButton, BOOL* isRewarded) {
     if (delegateObject == nil)
         delegateObject = [[CrossDKBridge alloc] init];

     [delegateObject displayOverlayWithFormat:format position:position withCloseButton:withCloseButton isRewarded:isRewarded];
 }

@implementation CrossDKOverlayDelegate

-(instancetype)initWithParentBridge:(CrossDKBridge *)parentBridge {
    self = [super init];
    if ( self ) {
        self.parentBridge = parentBridge;
    }
    return self;
}

-(void)overlayWillStartPresentation {
    unitySendMessage("OverlayWillStartPresentation", @"Overlay will start presentation");
}

-(void)overlayDidFinishPresentation {
    unitySendMessage("OverlayDidFinishPresentation", @"Overlay did finish presentation");
}

-(void)overlayWillStartDismissal {
    unitySendMessage("OverlayWillStartDismissal", @"Overlay will start dismissal");
}

-(void)overlayDidFinishDismissal {
    unitySendMessage("OverlayDidFinishDismissal", @"Overlay did finish dismissal");
}

-(void)overlayStartsPlayingVideo {
    unitySendMessage("OverlayStartsPlayingVideo", @"Overlay starts playing video");
}

-(void)overlayPlayedHalfVideo {
    unitySendMessage("OverlayPlayedHalfVideo", @"Overlay played half video");
}

-(void)overlayDidFinishPlayingVideo {
    unitySendMessage("OverlayDidFinishPlayingVideo", @"Video overlay did finish playing video");
}

-(void)overlayShowsRecommendedAppInAppStore {
    unitySendMessage("OverlayShowsRecommendedAppInAppStore", @"Overlay shows recommended app in AppStore");
}

-(void)overlayDidRewardUserWithReward {
    unitySendMessage("OverlayDidRewardUserWithReward", @"Overlay did reward user with reward");
}

-(void)overlayDidFailToLoadWithError:(NSError * _Nonnull)error {
    unitySendMessage("OverlayDidFailToLoadWithError", [NSString stringWithFormat:@"Overlay did fail to load with error : %ld", (long)error]);
}

-(void)overlayUnavailableWithError:(enum OverlayError)error {
    switch(error) {
        case OverlayErrorUnsupportedOSVersion:
            unitySendMessage("OverlayUnavailableWithError", @"Overlay error: unsupported iOS Version");
            break;
        case OverlayErrorUnavailableWindowScene:
            unitySendMessage("OverlayUnavailableWithError", @"Overlay error: unavailable window scene");
            break;
        case OverlayErrorUnavailableRecommendation:
            unitySendMessage("OverlayUnavailableWithError", @"Overlay error: unavailable recommendation");
            break;
        case OverlayErrorNoConfiguration:
            unitySendMessage("OverlayUnavailableWithError", @"Overlay error: no configuration");
            break;
    }
}

@end

File: Assets/Plugins/Android/CrossDKBridge.kt
package com.adikteev.unityadapter

import android.app.Activity
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.adikteev.crossdk.CrossDKConfig
import com.adikteev.crossdk.network.OnInitCrossDKListener
import com.adikteev.crossdk.views.CrossDKInterstitialView
import com.adikteev.crossdk.views.CrossDKMidSizeView
import com.adikteev.crossdk.views.CrossDKView
import com.adikteev.crossdk.views.listener.CrossDKContentCallback
import com.adikteev.crossdk.views.listener.CrossDKLoadCallback
import com.adikteev.crossdk.views.listener.CrossDKRewardedCallback
import com.adikteev.crossdk.views.position.CrossDKPosition
import com.unity3d.player.UnityPlayer


/*
**CrossDKBridge for showing CrossDK android format
 */

public class CrossDKBridge {
    private var mUnityPlayerActivity: Activity = UnityPlayer.currentActivity
    private var mOpenedOverlayFormat: OverlayFormat = OverlayFormat.NONE
    private lateinit var mCrossDKView: CrossDKView
    private lateinit var mCrossDKMidSizeView: CrossDKMidSizeView
    private lateinit var mCrossDKInterstitialView: CrossDKInterstitialView

    ///////////////////////////////////////////////////////////////////////////
    // CONFIG
    ///////////////////////////////////////////////////////////////////////////

    public fun config(appId: String, apiKey: String, userId: String?, deviceId:String?) {
        CrossDKConfig.Builder()
            .apiKey(apiKey)
            .appId(appId)
            .userId(userId)
            .deviceId(deviceId)
            .setup(mUnityPlayerActivity, object : OnInitCrossDKListener {
                override fun onInitSuccess() {}
                override fun onInitFailure(exception: Exception?) {
                    unitySendOverlayError("Overlay error: configuration error")
                }
            })
    }

    ///////////////////////////////////////////////////////////////////////////
    // DISPLAY
    ///////////////////////////////////////////////////////////////////////////

    public fun displayOverlay(
        format: Int,
        position: Int,
        isCloseButtonVisible: Boolean,
        isRewarded: Boolean
    ) {
        val crossDKContentCallback = object : CrossDKContentCallback {
            override fun onConfigurationError() {
                unitySendOverlayError("Overlay error: configuration error")
            }

            override fun onNoRecommendation() {
                unitySendOverlayError("Overlay error: unavailable recommendation")
            }

            override fun onShowContentError() {
                unitySendOverlayError("Overlay error: show content error")
            }

            override fun onRecommendationDisplayed() {
                overlayDidFinishPresentation()
            }

            override fun onRecommendationClicked() {
                overlayShowsRecommendedAppInAppStore()
            }

            override fun onRecommendationClosed() {
                overlayDidFinishDismissal()
                destroyViews()
            }

            override fun onUnsupportedApiVersion() {
                unitySendOverlayError("Overlay error: unsupported Api version")
            }
        }
        mUnityPlayerActivity.runOnUiThread {
            if (mOpenedOverlayFormat != OverlayFormat.NONE) dismissOverlay()
            val overlayFormat = OverlayFormat.fromInt(format)
            mOpenedOverlayFormat = overlayFormat
            overlayWillStartPresentation()
            when (overlayFormat) {
                OverlayFormat.BANNER -> {
                    mCrossDKView = CrossDKView(mUnityPlayerActivity)
                    mCrossDKView.setCrossDKContentCallback(crossDKContentCallback)
                    mCrossDKView.setCloseButtonVisibility(if (isCloseButtonVisible) View.VISIBLE else View.INVISIBLE)
                    mUnityPlayerActivity.addContentView(mCrossDKView, getLayoutParams())
                    mCrossDKView.setPosition(if (position == 0) CrossDKPosition.BOTTOM else CrossDKPosition.BOTTOM_RAISED)
                }
                OverlayFormat.MID_SIZE -> {
                    mCrossDKMidSizeView = CrossDKMidSizeView(mUnityPlayerActivity)
                    mCrossDKMidSizeView.setCrossDKContentCallback(crossDKContentCallback)
                    mCrossDKMidSizeView.setCloseButtonVisibility(if (isCloseButtonVisible) View.VISIBLE else View.INVISIBLE)
                    mUnityPlayerActivity.addContentView(mCrossDKMidSizeView, getLayoutParams())
                    mCrossDKMidSizeView.setPosition(if (position == 0) CrossDKPosition.BOTTOM else CrossDKPosition.BOTTOM_RAISED)
                }
                OverlayFormat.INTERSTITIAL -> {
                    mCrossDKInterstitialView = CrossDKInterstitialView(mUnityPlayerActivity)
                    mCrossDKInterstitialView.setCrossDKContentCallback(crossDKContentCallback)
                    mCrossDKInterstitialView.setCloseButtonVisibility(if (isCloseButtonVisible) View.VISIBLE else View.INVISIBLE)
                    mCrossDKInterstitialView.setRewarded(isRewarded,
                        object : CrossDKRewardedCallback {
                            override fun onUserRewarded() {
                                overlayDidFinishPlayingVideo()
                                overlayDidRewardUserWithReward()
                            }
                        })
                    mUnityPlayerActivity.addContentView(mCrossDKInterstitialView, getLayoutParams())
                    mCrossDKInterstitialView.load(object : CrossDKLoadCallback {
                        override fun onRecommendationLoaded() {
                            mCrossDKInterstitialView.show()
                        }

                        override fun onRecommendationLoadFailure() {
                            overlayDidFailToLoadWithError(java.lang.Exception("Recommendation load failure"))
                        }
                    })
                }
                else -> {
                    unitySendOverlayError("Overlay error: unsupported format requested")
                }
            }
        }
    }

    public fun dismissOverlay() {
        overlayWillStartDismissal()
        mUnityPlayerActivity.runOnUiThread {
            when (mOpenedOverlayFormat) {
                OverlayFormat.BANNER -> {
                    mCrossDKView.dismissView(true)
                    destroyViews()
                }
                OverlayFormat.MID_SIZE -> {
                    mCrossDKMidSizeView.dismissView(true)
                    destroyViews()
                }
                OverlayFormat.INTERSTITIAL -> {
                    mCrossDKInterstitialView.dismissView(true)
                    destroyViews()
                }
                else -> {}
            }
        }
    }

    public fun destroyViews() {
        mUnityPlayerActivity.runOnUiThread {
            when (mOpenedOverlayFormat) {
                OverlayFormat.BANNER -> {
                    mCrossDKView.destroy()
                    val parentView = mCrossDKView.parent
                    if (parentView is ViewGroup) {
                        parentView.removeView(mCrossDKView)
                    }
                }
                OverlayFormat.MID_SIZE -> {
                    mCrossDKMidSizeView.destroy()
                    val parentView = mCrossDKMidSizeView.parent
                    if (parentView is ViewGroup) {
                        parentView.removeView(mCrossDKMidSizeView)
                    }
                }
                OverlayFormat.INTERSTITIAL -> {
                    mCrossDKInterstitialView.destroy()
                    val parentView = mCrossDKInterstitialView.parent
                    if (parentView is ViewGroup) {
                        parentView.removeView(mCrossDKInterstitialView)
                    }
                }
                else -> {}
            }
            mOpenedOverlayFormat = OverlayFormat.NONE
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // LISTENERS
    ///////////////////////////////////////////////////////////////////////////

    private fun overlayWillStartPresentation() {
        unitySendMessage("OverlayWillStartPresentation", "Overlay will start presentation");
    }

    private fun overlayDidFinishPresentation() {
        unitySendMessage("OverlayDidFinishPresentation", "Overlay did finish presentation");
    }

    private fun overlayWillStartDismissal() {
        unitySendMessage("OverlayWillStartDismissal", "Overlay will start dismissal");
    }

    private fun overlayDidFinishDismissal() {
        unitySendMessage("OverlayDidFinishDismissal", "Overlay did finish dismissal");
    }

    private fun overlayStartsPlayingVideo() {
        unitySendMessage("OverlayStartsPlayingVideo", "Overlay starts playing video");
    }

    private fun overlayPlayedHalfVideo() {
        unitySendMessage("OverlayPlayedHalfVideo", "Overlay played half video");
    }

    private fun overlayDidFinishPlayingVideo() {
        unitySendMessage("OverlayDidFinishPlayingVideo", "Video overlay did finish playing video");
    }

    private fun overlayShowsRecommendedAppInAppStore() {
        unitySendMessage(
            "OverlayShowsRecommendedAppInAppStore",
            "Overlay shows recommended app in AppStore"
        );
    }

    private fun overlayDidRewardUserWithReward() {
        unitySendMessage("OverlayDidRewardUserWithReward", "Overlay did reward user with reward");
    }

    private fun overlayDidFailToLoadWithError(exception: Exception) {
        unitySendMessage(
            "OverlayDidFailToLoadWithError",
            "Overlay did fail to load with error: ${exception.message}"
        )
    }

    ///////////////////////////////////////////////////////////////////////////
    // HELPERS
    ///////////////////////////////////////////////////////////////////////////

    private fun getLayoutParams(): FrameLayout.LayoutParams {
        val adParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT
        )
        adParams.gravity = Gravity.BOTTOM
        return adParams
    }

    private fun unitySendOverlayError(message: String) {
        unitySendMessage("OverlayUnavailableWithError", message)
    }

    private fun unitySendMessage(method: String, message: String) {
        UnityPlayer.UnitySendMessage("CrossDK", method, message)
    }

    private enum class OverlayFormat(val value: Int) {
        NONE(-1),
        BANNER(0),
        MID_SIZE(1),
        INTERSTITIAL(2);

        companion object {
            fun fromInt(value: Int) = values().first { it.value == value }
        }
    }
}